net/dns/nrpt_windows.go

```diff
@@ -0,0 +1,297 @@
+			db.logf("Error deleting NRPT rule key %q: %v", keyNameLocal, err)
+			return err
+		}
+
+		keyNameGP := nrptBaseGP + `\` + rid
+		err := registry.DeleteKey(registry.LOCAL_MACHINE, keyNameGP)
+		if err == nil {
+			// If this deleted subkey existed under the GP key, we will need to refresh.
+			db.isGPDirty = true
+		} else if err != registry.ErrNotExist {
+			db.logf("Error deleting NRPT rule key %q: %v", keyNameGP, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (db *nrptRuleDatabase) WriteSplitDNSConfig(servers []string, domains []dnsname.FQDN) error {
+	// NRPT has an undocumented restriction that each rule may only be associated
+	// with a maximum of 50 domains. If we are setting rules for more domains
+	// than that, we need to split domains into chunks and write out a rule per chunk.
+	dq := len(domains) / nrptMaxDomainsPerRule
+	dr := len(domains) % nrptMaxDomainsPerRule
+
+	domainRulesLen := dq
+	if dr > 0 {
+		domainRulesLen++
+	}
+
+	db.loadRuleSubkeyNames()
+	for len(db.ruleIDs) < domainRulesLen {
+		guid, err := windows.GenerateGUID()
+		if err != nil {
+			return err
+		}
+		db.ruleIDs = append(db.ruleIDs, guid.String())
+	}
+
+	// Remove any surplus rules that are no longer needed.
+	ruleIDsToRemove := db.ruleIDs[domainRulesLen:]
+	db.delRuleKeys(ruleIDsToRemove)
+
+	// We need to save the list of rule IDs to our Tailscale registry key so that
+	// we know which rules are ours during subsequent modifications to NRPT rules.
+	ruleIDsToWrite := db.ruleIDs[:domainRulesLen]
+	if len(ruleIDsToWrite) == 0 {
+		if err := winutil.DeleteRegValue(nrptRuleIDValueName); err != nil {
+			return err
+		}
+		db.ruleIDs = nil
+		return nil
+	}
+
+	if err := winutil.SetRegStrings(nrptRuleIDValueName, ruleIDsToWrite); err != nil {
+		return err
+	}
+	db.ruleIDs = ruleIDsToWrite
+
+	doms := make([]string, 0, nrptMaxDomainsPerRule)
+	for i := 0; i < domainRulesLen; i++ {
+		// Each iteration consumes nrptMaxDomainsPerRule domains...
+		curLen := nrptMaxDomainsPerRule
+		// Except for the final iteration: when we have a remainder, use that instead.
+		if i == domainRulesLen-1 && dr > 0 {
+			curLen = dr
+		}
+
+		// Obtain the slice of domains to consume within the current iteration.
+		start := i * nrptMaxDomainsPerRule
+		end := start + curLen
+		for _, domain := range domains[start:end] {
+			// NRPT rules must have a leading dot, which is not usual for
+			// DNS search paths.
+			doms = append(doms, "."+domain.WithoutTrailingDot())
+		}
+
+		if err := db.writeNRPTRule(db.ruleIDs[i], servers, doms); err != nil {
+			return err
+		}
+
+		doms = doms[:0]
+	}
+
+	return nil
+}
+
+// Refresh notifies the Windows group policy engine when policies have changed.
+func (db *nrptRuleDatabase) Refresh() {
+	if !db.isGPDirty {
+		return
+	}
+	ok, _, err := procRefreshPolicyEx.Call(
+		uintptr(1), // Refresh computer policy, not user policy
+		uintptr(_RP_FORCE),
```

Comment:
Rather than forcing, is there a way that we could actually create a real policy that Windows loads and applies? It seems like that would ensure all the correct mechanisms get tickled along the way, and perhaps would make cleanup easier too.

Reply:
Unfortunately the changes aren't being seen without the force flag, so for now at least I think we need to keep it.